### ISSUE TITLE
Error installing on Ubuntu 20.04

### ISSUE THREAD

**Issue by Ziaeemehr** [fastlib/fCWT#1]
Hello and thank you for sharing the package.
I just edited the CMakelist line 43 as follows:
```
set(CMAKE_C_COMPILER "gcc-9")
set(CMAKE_CXX_COMPILER "g++-9")
```
and I got the following error for making :

```
[ 12%] Building CXX object CMakeFiles/fCWT_example.dir/src/fcwt.cpp.o
[ 25%] Building CXX object CMakeFiles/fCWT_example.dir/src/main.cpp.o
[ 37%] Building CXX object CMakeFiles/fCWT_example.dir/src/rwave.cpp.o
[ 50%] Building CXX object CMakeFiles/fCWT_example.dir/src/wavelib.cpp.o
[ 62%] Linking CXX executable fCWT_example
/usr/bin/ld: CMakeFiles/fCWT_example.dir/src/fcwt.cpp.o: in function `fcwt::create_optimization_schemes(int, int, int)':
fcwt.cpp:(.text+0x303): undefined reference to `fftwf_init_threads'
/usr/bin/ld: fcwt.cpp:(.text+0x30f): undefined reference to `fftwf_plan_with_nthreads'
/usr/bin/ld: fcwt.cpp:(.text+0x3a2): undefined reference to `fftwf_plan_dft_r2c_1d'
/usr/bin/ld: fcwt.cpp:(.text+0x418): undefined reference to `fftwf_plan_dft_1d'
/usr/bin/ld: fcwt.cpp:(.text+0x420): undefined reference to `fftwf_export_wisdom_to_filename'
/usr/bin/ld: fcwt.cpp:(.text+0x430): undefined reference to `fftwf_free'
/usr/bin/ld: fcwt.cpp:(.text+0x438): undefined reference to `fftwf_free'
/usr/bin/ld: fcwt.cpp:(.text+0x4e3): undefined reference to `fftwf_alloc_complex'
/usr/bin/ld: fcwt.cpp:(.text+0x4ee): undefined reference to `fftwf_alloc_complex'
/usr/bin/ld: CMakeFiles/fCWT_example.dir/src/fcwt.cpp.o: in function `fcwt::load_optimization_schemes(bool, int, int)':
fcwt.cpp:(.text+0x76e): undefined reference to `fftwf_import_wisdom_from_filename'
/usr/bin/ld: CMakeFiles/fCWT_example.dir/src/fcwt.cpp.o: in function `fcwt::main(float*, float*, int*, int*, int*, int*, float*, int, bool)':
fcwt.cpp:(.text.startup+0xcf): undefined reference to `fftwf_alloc_complex'
/usr/bin/ld: fcwt.cpp:(.text.startup+0xdf): undefined reference to `fftwf_alloc_complex'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x139): undefined reference to `fftwf_init_threads'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x145): undefined reference to `fftwf_plan_with_nthreads'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x16a): undefined reference to `fftwf_plan_dft_r2c_1d'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x176): undefined reference to `fftwf_execute'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x17f): undefined reference to `fftwf_destroy_plan'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x198): undefined reference to `fftwf_plan_dft_1d'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x2e1): undefined reference to `fftwf_execute_dft'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x329): undefined reference to `fftwf_destroy_plan'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x331): undefined reference to `fftwf_free'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x339): undefined reference to `fftwf_free'
/usr/bin/ld: fcwt.cpp:(.text.startup+0x33e): undefined reference to `fftwf_cleanup_threads'
/usr/bin/ld: CMakeFiles/fCWT_example.dir/src/wavelib.cpp.o: in function `wavelib::cwt(double*, int, int, int)':
wavelib.cpp:(.text+0x2e): undefined reference to `cwt_init'
/usr/bin/ld: wavelib.cpp:(.text+0x61): undefined reference to `setCWTScales'
/usr/bin/ld: wavelib.cpp:(.text+0x6c): undefined reference to `cwt'
/usr/bin/ld: wavelib.cpp:(.text+0x78): undefined reference to `cwt_free'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/fCWT_example.dir/build.make:132: fCWT_example] Error 1
make[1]: *** [CMakeFiles/Makefile2:106: CMakeFiles/fCWT_example.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

```
g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Do I need to use gcc 10?

**Comment by fastlib** [comment]
Dear @Ziaeemehr, thank you for reaching out! Looking at your output, I notice that FFTW is not properly linked which causes the majority if not all errors. This should not be caused by the difference in gcc compiler. It could be caused by the fact I built FFTW's library on a Mac which cannot be used on Linux. I cannot test this as I only have the possibility to test on Mac OSX and Windows. It could be that FFTW must be built and configured for Linux system specifically. Can you check this? If this works and you built the static libraries, you can do a pull-request so we can fix this for all Linux users.

More info about FFTW and its library can be found at their wonderful documentation https://www.fftw.org/fftw3_doc/Installation-on-Unix.html#Installation-on-Unix 

Let me know if you have any questions!

**Comment by Ziaeemehr** [comment]
I tried it on both Macbook 2020 and Ubuntu 20.04, both have the same error,
I can also link fftw3 for compiling a C++ code, so g++ can find it using something like this:
```sh
g++ -std=c++11 -O2 my_code.cpp -o prog -lm -lfftw3
```
I installed fftw3 by `brew` in mac and `sudo apt-get` in Ubunto, but I think it does not make difference as long as the compiler (`gcc` or `g++`) can find it.

I am not sure in this case but these errors usually appear when I forget to add `-lfftw3`.

**Comment by fastlib** [comment]
Alright, good to know the error is persistent across systems. I see you provided your own FFTW installation. I expect that's where things go. We link FFTW in the src/fftw folder via the CMake 'find_library()' function. This function is configured such that it searches in the src/fftw folder. 

I imagine linking is going wrong when you provide FFTW yourself. You could try to configure the search path in cmakelists.txt at line 121, and line 129 for the OMP version to search for your HomeBrew FFTW version (I expect in the /usr/local/lib folder). Make sure you have both the default FFTW library built as well as a multi-threaded version (omp is fastest, but pthreads is also possible). 

Let me know if this solves the linking problem.

**Comment by maedoc** [comment]
The missing symbols there are for the 32-bit float version of the library, this should be installed and linked explicitly in some cases.

**Comment by maedoc** [comment]
Sorry, I missed the part where your CMake links the macOS object files regardless of platform.  That's pretty weird.  In any case it builds on my Mac.

**Comment by maedoc** [comment]
@Ziaeemehr try removing the fftw3f.a files in the fCWT repo, `apt-get install -y libfftw3-single3` and check that after `cmake` it reports that the FFTW3 library found is in `/usr/...` not the fCWT folder.

**Comment by Ziaeemehr** [comment]
modifing the CMakelist as follows and remowing library files from `src/fftw` resolved the errors from fftw 
```bash
find_library (
    FFTW
    NAMES fftw3f libfftw3f # what to look for
    # HINTS "${PROJECT_SOURCE_DIR}/src/fftw" # where to look
    # NO_DEFAULT_PATH # do not search system default paths
    REQUIRED
)
if(UNIX)
	find_library (
		FFTW_OMP
		NAMES fftw3f_omp libfftw3f_omp # what to look for
		# HINTS "${PROJECT_SOURCE_DIR}/src/fftw" # where to look
		# NO_DEFAULT_PATH # do not search system default paths
		REQUIRED
	)
```



```sh
Build type is Release
Building of shared library is enabled.
/home/ziaee/prog/install_files/fCWT
STATUS OF LIBRARY: FFTW: [/usr/lib/x86_64-linux-gnu/libfftw3f.so]
STATUS OF LIBRARY: FFTW_OMP: [/usr/lib/x86_64-linux-gnu/libfftw3f_omp.so]
STATUS OF LIBRARY: WAVELIB: [/home/ziaee/prog/install_files/fCWT/src/libwavelib/libwavelib.a]
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ziaee/prog/install_files/fCWT/build
```

but doing the same thing for wavelib did not work. cmake still suggests `src/libwavelib/libwavelib.a`

**Comment by fastlib** [comment]
Good to hear that editing lines 121,122 and 129,130 in the CMakeLists file works! I don't see how wavelib fails to link. Did you also installed wavelib using HomeBrew or via another method?

**Comment by Ziaeemehr** [comment]
on mac the package seem to be installed correctly:

```bash
(base) ➜  build git:(main)  make
[ 14%] Building CXX object CMakeFiles/fCWT.dir/src/fcwt.cpp.o
[ 28%] Linking CXX shared library libfCWT.dylib
[ 28%] Built target fCWT
[ 42%] Building CXX object CMakeFiles/fCWT_example.dir/src/fcwt.cpp.o
[ 57%] Building CXX object CMakeFiles/fCWT_example.dir/src/main.cpp.o
[ 71%] Building CXX object CMakeFiles/fCWT_example.dir/src/rwave.cpp.o
[ 85%] Building CXX object CMakeFiles/fCWT_example.dir/src/wavelib.cpp.o
[100%] Linking CXX executable fCWT_example
[100%] Built target fCWT_example
(base) ➜  build git:(main) ✗ sudo make install
Password:
sudo: a password is required
(base) ➜  build git:(main)  sudo make install
Password:
Consolidate compiler generated dependencies of target fCWT
[ 28%] Built target fCWT
Consolidate compiler generated dependencies of target fCWT_example
[100%] Built target fCWT_example
Install the project...
-- Install configuration: "Release"
-- Installing: /usr/local/include/fcwt.h
-- Installing: /usr/local/lib/libfCWT.1.0.dylib
/usr/local/opt/llvm/bin/install_name_tool: error: unsupported load command (cmd=0x80000034)
-- Installing: /usr/local/lib/libfCWT.dylib
-- Installing: /usr/local/share/fcwt/cmake/fCWT-config.cmake
-- Installing: /usr/local/share/fcwt/cmake/fCWT-config-release.cmake
```

on Ubuntu:
I don't have other installations of `wavelib`, probably need to be built by adding it to the project.

**Comment by fastlib** [comment]
For Ubuntu, you have two options:
1. Remove `wavelib` from the example and compile without `wavelib` as it is only used for benchmark purposes, or
2. you could rebuild `wavelib`'s static library on Ubuntu using their source-code https://github.com/rafat/wavelib 

Let me know if you succeed!

**Comment by Ziaeemehr** [comment]
for Ubuntu to build the static library for wavelib:
```bash
git clone git@github.com:rafat/wavelib.git
cd wavelib
mkdir -p build
cd build
cmake ../
cp Bin/* /copy_to_directory_of_installation_fCWT/src/libwavelib/
```

then continue with the installation of `fCWT` 
```
cd fCWT/build
cmake ..
```

now all the dependencies are found:

```bash
Build type is Release
Building of shared library is enabled.
/home/ziaee/prog/install_files/fCWT
STATUS OF LIBRARY: FFTW: [/usr/lib/x86_64-linux-gnu/libfftw3f.so]
STATUS OF LIBRARY: FFTW_OMP: [/usr/lib/x86_64-linux-gnu/libfftw3f_omp.so]
STATUS OF LIBRARY: WAVELIB: [/home/ziaee/prog/install_files/fCWT/src/libwavelib/libwavelib.a]
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ziaee/prog/install_files/fCWT/build
```

```
$ make
$ sudo make install
```
```bash
[ 62%] Built target fCWT_example
[100%] Built target fCWT
Linking CXX shared library CMakeFiles/CMakeRelink.dir/libfCWT.so
Install the project...
-- Install configuration: "Release"
-- Installing: /usr/local/include/fcwt.h
-- Installing: /usr/local/lib/libfCWT.so.1.0
-- Installing: /usr/local/lib/libfCWT.so
-- Installing: /usr/local/share/fcwt/cmake/fCWT-config.cmake
-- Installing: /usr/local/share/fcwt/cmake/fCWT-config-release.cmake
```

**Comment by sitadrost** [comment]
I seem to be running into the same problem on Ubuntu 20.04 (with gcc-10): upon my first attempt to build fCWT I got these "undefined reference" errors regarding FFTW, so I was delighted when I found this post.
Following the instructions on their website, I then installed FFTW, and edited lines 121, 122, 129 and 130 in `CMakeLists.txt`.
Unfortunately, CMake still can't find the required FFTW lib files, even when I add a hint pointing to `/usr/lib/x86_64-linux-gnu`
FFTW lib files in this folder:
```
libfftw3f_omp.so.3
libfftw3f_omp.so.3.5.8
libfftw3f.so.3
libfftw3f.so.3.5.8
libfftw3f_threads.so.3
libfftw3f_threads.so.3.5.8
libfftw3_omp.so.3
libfftw3_omp.so.3.5.8
libfftw3.so.3
libfftw3.so.3.5.8
libfftw3_threads.so.3
libfftw3_threads.so.3.5.8

```
Wrong version? Something else I'm doing wrong?
Thanks in advance!
